### PR TITLE
[SM-628] Add trim validator to SM dialogs

### PR DESCRIPTION
**Pull request by Thomas-Avery**
## Type of change

```
- [X] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

The purpose of this PR is to add a validator to SM dialogs that prevents names that start or end with whitespace.

## Code changes

- apps/web/src/locales/en/messages.json

Added new error message for input validator

- bitwarden_license/bit-web/src/app/secrets-manager/projects/dialog/project-dialog.component.ts
bitwarden_license/bit-web/src/app/secrets-manager/secrets/dialog/secret-dialog.component.ts
bitwarden_license/bit-web/src/app/secrets-manager/service-accounts/access/dialogs/access-token-create-dialog.component.ts
bitwarden_license/bit-web/src/app/secrets-manager/service-accounts/dialog/service-account-dialog.component.ts

Adding in the new validator.

To note, I purposely left out the value field for secrets.
There might be a use case for secret values to start or end with whitespace? 

- libs/components/src/form-field/bit-validators/index.ts
libs/components/src/form-field/bit-validators/trim.validator.spec.ts
libs/components/src/form-field/bit-validators/trim.validator.ts

Code and unit tests for the new validator.

- libs/components/src/form-field/error.component.ts

Adding validator to the error.component.

- libs/components/src/form-field/bit-validators.stories.ts

Adding a story showcasing the custom validator.

### Secret

https://user-images.githubusercontent.com/43214426/224845947-4abade35-5139-45fa-b193-481090022846.mov

### Project

https://user-images.githubusercontent.com/43214426/224846021-84cf06d3-02e8-4be8-a420-82d12d379252.mov

### Service Account

https://user-images.githubusercontent.com/43214426/224846064-29b12adf-a106-47d6-b21b-95b7daad2987.mov

### Access token

https://user-images.githubusercontent.com/43214426/224846092-330d762f-f0a5-4559-894e-0eeeb70fc5c4.mov

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

### CODE REVIEW

**Review by willmartian — APPROVED** [review]
Looks great! 🥳

**Review by Hinton — COMMENTED** [review]
Do we want to make this a generic validator, and add the i18n logic to `error.component`?

**Comment by Hinton** [comment]
@Thomas-Avery maybe but it seems generic enough that not having to duplicate the string everywhere feels better than making it flexible. We can always add an optional message at a later stage.

**Comment by Thomas-Avery** [comment]
> @Thomas-Avery maybe but it seems generic enough that not having to duplicate the string everywhere feels better than making it flexible. We can always add an optional message at a later stage.

Okay should be swapped to generic implementation now.

File: libs/components/src/form-field/bit-validators/trim.validator.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    if (value != value.trim()) {
+      return {
+        trim: {
+          message: "input is not trimmed",
+        },
+      };
+    }
+    return null;
+  };
+}
```
Comment (Hinton) [inline libs/components/src/form-field/bit-validators/trim.validator.ts:22]: We don't need an argument, so we should be able to rewrite this to something like:

```suggestion
export function trimValidator(control: AbstractControl): ValidationErrors | null => {
  if (!(control instanceof FormControl)) {
    throw new Error("trimValidator only supports validating FormControls");
  }

  if (control.value === null || control.value === undefined) {
    return null;
  }
  const value = control.value;
  if (value != value.trim()) {
    return {
      trim: {
        message: "input is not trimmed",
      },
    };
  }
  return null;
};
```

Which lets us add the validator with just `BitValidators.trimValidator`

Comment (Thomas-Avery) [reply libs/components/src/form-field/bit-validators/trim.validator.ts]: Resolved in https://github.com/bitwarden/clients/pull/4993/commits/3a13bf23eadd4f45c400c37002e8486fa8a0986d

**Review by Hinton — DISMISSED** [review]

**Comment by willmartian** [comment]
Based on feedback from design, I updated the validator to automatically trim whitespace. Now, the user will only see a validation error when entering a value that is entirely whitespace. 

As such, I also updated the validation message to read as `"Input must not contain only whitespace."`

**Review by Thomas-Avery — COMMENTED** [review]
Thank you, @willmartian this looks great.

Someone else on the team will need to approve.

**Review by willmartian — APPROVED** [review]

**Comment by jtouchstonebw** [comment]
Overall, this looks good! One thing I noticed was that the (X) icon in the error message appeared a few milliseconds after the text appeared. Not a deal breaker, but ideally they would load together at the same time so that text isn't moving around when someone is trying to read it.

**Review by willmartian — DISMISSED** [review]

**Comment by willmartian** [comment]
Updated the validator to run on `'submit'` instead of `'blur'`, due to it not working when submitting a form with the `enter` button.

**Review by Thomas-Avery — COMMENTED** [review]
Changes to update on submit looks good.

**Review by willmartian — DISMISSED** [review]

**Review by willmartian — APPROVED** [review]

**Pull request merged**